cmd/otel-allocator/allocation/allocator.go

```diff
@@ -0,0 +1,131 @@
+package allocation
+
+import (
+	"log"
+	"sync"
+)
+
+/*
+	Load balancer will serve on an HTTP server exposing /jobs/<job_id>/targets <- these are configured using least connection
+	Load balancer will need information about the collectors in order to set the URLs
+	Keep a Map of what each collector currently holds and update it based on new scrape target updates
+*/
+
+// Create a struct that holds collector - and jobs for that collector
+// This struct will be parsed into endpoint with collector and jobs info
+
+type collector struct {
+	Name       string
+	NumTargets int
+}
+
+// Allocator makes decisions to distribute work among
+// a number of OpenTelemetry collectors based on the number of targets.
+// Users need to call SetTargets when they have new targets in their
+// clusters and call Reshard to process the new targets and reshard.
+type Allocator struct {
+	m sync.Mutex
+
+	targetsWaiting map[string]TargetItem // temp buffer to keep targets that are waiting to be processed
+
+	collectors map[string]*collector // all current collectors
+
+	nextCollector *collector
+	targetItems   map[string]*TargetItem
+}
+
+// findNextCollector finds the next collector with less number of targets.
+func (allocator *Allocator) findNextCollector() {
+	for _, v := range allocator.collectors {
+		if v.NumTargets < allocator.nextCollector.NumTargets {
```

Comment:
If `SetCollectors` is called with 0 collectors, it would not enter the loop itself since _allocator.collectors_ will not have any elements thus ensuring `nextCollector` is never reached.